quinn-proto/src/endpoint.rs

```diff
@@ -499,15 +499,26 @@ impl Endpoint {
         let (retry_src_cid, orig_dst_cid) = if header.token.is_empty() {
             (None, header.dst_cid)
         } else {
-            match Token::decode(&*server_config.token_key, &addresses.remote, &header.token) {
-                Ok(token)
-                    if token.inner.issued + server_config.retry_token_lifetime
-                        > SystemTime::now() =>
-                {
-                    (Some(header.dst_cid), token.inner.orig_dst_cid)
+            let valid_token =
+                Token::decode(&*server_config.token_key, &addresses.remote, &header.token)
+                    .and_then(|token| {
+                        let TokenInner {
+                            orig_dst_cid,
+                            issued,
+                        } = token.inner;
+                        if issued + server_config.retry_token_lifetime > SystemTime::now() {
+                            Ok((Some(header.dst_cid), orig_dst_cid))
+                        } else {
+                            Err(TokenDecodeError::InvalidRetry)
+                        }
+                    });
```

Comment:
Feels weird to use an `and_then()` combinator on a case where the closure is far larger than the original expression. Maybe this should be a method on `Token` instead?

Reply:
Done